src/pd/retry.rs

```diff
@@ -108,24 +115,84 @@ impl fmt::Debug for RetryClient {
     }
 }
 
+// A helper struct to check whether a previous connecting is running.
+struct ShouldReconnect<'a> {
+    connected: MutexLockFuture<'a, bool>,
+    first_ready: Option<bool>,
+}
+
+impl<'a> ShouldReconnect<'a> {
+    fn new(connected: MutexLockFuture<'a, bool>) -> Self {
+        ShouldReconnect {
+            connected,
+            first_ready: None,
+        }
+    }
+}
+
+impl<'a> Future for ShouldReconnect<'a> {
+    // `tuple.0 : MutexGuard` tells whether current connecting is finished, `tuple.1 : bool` tells if
+    // we should start a new connection.
+    type Output = (MutexGuard<'a, bool>, bool);
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match Pin::new(&mut self.connected).poll(cx) {
+            Poll::Ready(val) => match self.first_ready {
+                None => {
+                    self.first_ready = Some(true);
+                    Poll::Ready((val, true))
+                }
+                Some(true) => panic!("should not happen"),
+                Some(false) => {
+                    let res = !*val;
+                    Poll::Ready((val, res))
+                }
+            },
+            Poll::Pending => {
+                if self.first_ready.is_none() {
+                    self.first_ready = Some(false)
+                }
+                Poll::Pending
+            }
+        }
+    }
+}
+
 // A node-like thing that can be connected to.
+#[async_trait]
 trait Reconnect {
     type Cl;
-    fn reconnect(&self, interval: u64) -> Result<()>;
+    async fn reconnect(&self, interval: u64) -> Result<()>;
     fn with_cluster<T, F: Fn(&Self::Cl) -> T>(&self, f: F) -> T;
 }
 
+#[async_trait]
 impl Reconnect for RetryClient<Cluster> {
     type Cl = Cluster;
 
-    fn reconnect(&self, interval: u64) -> Result<()> {
-        if let Some(cluster) =
-            self.connection
-                .reconnect(&self.cluster.read().unwrap(), interval, self.timeout)?
-        {
-            *self.cluster.write().unwrap() = cluster;
+    async fn reconnect(&self, interval: u64) -> Result<()> {
+        let connected: MutexLockFuture<bool> = self.connected.lock();
+        let (mut connected, should_connect) = ShouldReconnect::new(connected).await;
+        // if try to reconnect while a previous reconnect is running,
+        // the second reconnect will succeed in direct if the first one is succeed
+        if should_connect {
+            *connected = false;
```

Comment:
I'm a bit confused about this, since we hold the lock on `self.connected`, does that mean that this not observable to anyone else? I.e., there is no point in setting this to false then true?

Reply:
> I'm a bit confused about this, since we hold the lock on `self.connected`, does that mean that this not observable to anyone else? I.e., there is no point in setting this to false then true?

Process may fail at Line 179 ~ Line 190,then value of connected will stay false, next request can start reconnect instantly.